src/dune/build_system.ml

```diff
@@ -1340,11 +1351,49 @@ end = struct
       | exception Unix.Unix_error (ENOENT, _, _) -> ()
       | () -> () )
 
+  let execute_action_until_all_deps_ready ~context ~env ~targets ~rule_loc
+    action =
+    let open Dune_action.Protocol in
+    let provided_dependencies_set = ref Dependency.Set.empty
+    and stages = ref Appendable_list.empty in
+    let rec loop () =
+      let* result =
+        Action_exec.exec ~context ~env ~targets ~rule_loc
+          ~provided_dependencies:!provided_dependencies_set action
+      in
+      match result with
+      | Done -> Fiber.return (Appendable_list.to_list !stages)
+      | Need_more_deps deps ->
+        let deps_to_build = Dependency.Map.values deps |> Dep.Set.of_list in
+        (stages := Appendable_list.(!stages @ singleton deps_to_build));
+        let* () = build_deps deps_to_build in
+        provided_dependencies_set :=
+          Dependency.Set.union !provided_dependencies_set
+            (Dependency.Set.of_keys deps);
+        loop ()
+    in
+    loop ()
+
+  let compute_rule_digest (rule : Internal_rule.t) ~deps ~action ~sandbox_mode
+    =
+    let targets_as_list = Path.Build.Set.to_list rule.targets in
+    let env = Internal_rule.effective_env rule in
+    let trace =
+      ( Dep.Set.trace deps ~sandbox_mode ~env ~eval_pred
+      , List.map targets_as_list ~f:(fun p -> Path.to_string (Path.build p))
+      , Option.map rule.context ~f:(fun c -> c.name)
+      , Action.for_shell action )
+    in
+    Digest.generic trace
+
+  let compute_dependencies_digest deps ~sandbox_mode ~env ~eval_pred =
+    Dep.Set.trace deps ~sandbox_mode ~env ~eval_pred |> Digest.generic
```

Comment:
It seems we haven't been following this practice in the past here, but in general it's a good idea to have a type annotation next to such unsafely-polymorphic functions like `Digest.generic`. 
Consider what happens if an argument is added to `Dep.Set.trace`. The program will compile, but the result will either crash or worse: produce nonsensical digest.

Reply:
Good point, thank you!